src/turbomind/models/llama/LlamaWeight.cc

```diff
@@ -127,21 +128,22 @@ void LlamaWeight::release()
 
 void LlamaWeight::to_device(const core::Device& device)
 {
-    core::ContextGuard guard = context();
-
-    auto to_device = [&](Tensor& x) -> Tensor {
-        auto tmp = std::exchange(x, empty_like(x, device));
-        Copy(tmp, x);
-        return tmp;
-    };
-
-    std::vector<Tensor> tmp_cpu_tensors;
+    TM_CHECK(device.type == kCPU || device.type == kDEVICE);
+    core::ContextGuard guard{stream_, alloca_, Allocator{kCPUpinned}};
 
     auto tensor_ptr_map = get_parameters();
     for (auto& [name, tensor_ptr] : tensor_ptr_map) {
-        auto tmp_tensor = to_device(*tensor_ptr);
-        if (tmp_tensor.device().type != kDEVICE) {
-            tmp_cpu_tensors.push_back(tmp_tensor);
+        if (pinned_weights_.find(name) == pinned_weights_.end()) {
+            pinned_weights_[name] = empty_like(*tensor_ptr, kCPUpinned);
+        }
+        auto& pinned = pinned_weights_.at(name);
+        if (device.type == kCPU) {
+            Copy(*tensor_ptr, pinned);
+            *tensor_ptr = pinned;
+        }
+        else {
+            *tensor_ptr = empty_like(*tensor_ptr, kDEVICE);
+            Copy(pinned, *tensor_ptr);
```

Comment:
Out of context, this code appears to permit the transfer of empty pinned weights to the device, which would be misleading. I suggest implementing the following:
```
if (device_type == kCPU) {
   // 1. create pinned_weights[name] if it is not found
   // 2. cpy weight from device to pinnned mem
}
else {
   //1. check the pinned_weights[name] exist
   //2. cpy weight from pinned mem to device
}
```